### DISCUSSION TITLE
RT-DETR fine tuning "No labels found" during training

### DISCUSSION THREAD

**Discussion by MacJavellana** [ultralytics/ultralytics#19058] [Q&A]
helloo I've a similar problem with issue#13361 but i wasn't able to follow or fix the problem. I'm hoping that i could be able to fix my specific problem with your help. I'm fairly new in running machine learning models and using it

here was my warning/error: "train: WARNING ⚠️ No labels found in E:\thesis models\RT-DETR-Pipeline\CamoCrops_2025\Images\Train.cache. See https://docs.ultralytics.com/datasets for dataset formatting
 guidance."
 
 
my best guess is that my path and format isn't properly set due to the fact that my main dataset was in voc and i had to convert it in coco format

 for your reference here is my file structure
E:/thesis models/RT-DETR-Pipeline/
├── CamoCrops_2025/
│   ├── Annotations/
│   │   ├── Train/
│   │   ├── Validation/
│   │   └── Test_new/
│   ├── Images/
│   │   ├── Train/
│   │   ├── Validation/
│   │   └── Test_new/
│   ├── labels/
│   │   ├── annotations_train.json
│   │   ├── annotations_validation.json
│   │   └── annotations_test_new.json
├── main.py
├── eval.py
├── convert_to_coco.py
└── camocrops.yaml

here is my files
``` main.py
import torch
from ultralytics import RTDETR

if __name__ == '__main__':
    print(f"CUDA available: {torch.cuda.is_available()}")
    print(f"GPU device name: {torch.cuda.get_device_name(0)}")
    print(f"PyTorch CUDA version: {torch.version.cuda}")

    # SETTINGS
    MODEL = 'rtdetr-l'  # Changed to lighter model
    NAME = 'CamoCrops_rtdetr_final'
    DATA = 'E:/thesis models/RT-DETR-Pipeline/camocrops.yaml'
    EPOCHS = 1
    BATCH = 4  # Reduced batch size for 4GB VRAM
    #IMGSZ = 256  # Reduced image size
    LR0 = 0.01
    LRF = 0.0001
    MOMENTUM = 0.9
    OPTIMIZER = 'SGD'
    SAVE_PERIOD = 1
    VAL = False

    # Initialize model with pretrained weights
    model = RTDETR(MODEL + '.pt')

    # Train the model
    train_results = model.train(
        data=DATA,
        epochs=EPOCHS,
        batch=BATCH,
        #imgsz=IMGSZ,
        save_period=SAVE_PERIOD,
        device=0,
        name=NAME,
        optimizer=OPTIMIZER,
        lr0=LR0,
        lrf=LRF,
        momentum=MOMENTUM,
        val=VAL
    )
```

```camocrops.yaml
# Dataset paths
path: E:/thesis models/RT-DETR-Pipeline/CamoCrops_2025
train: Images/Train
val: Images/Validation
test: Images/Test_new

# COCO format annotations
train_ann: labels/annotations_train.json
val_ann: labels/annotations_validation.json
test_ann: label/sannotations_test_new.json

# Dataset type
format: coco

# Classes
names:
  0: ampalaya
  1: atis
  2: avocado
  3: bell_pepper
  4: calamansi
  5: cucumber
  6: dalandan
  7: french_bean
  8: green_chili
  9: green_mango
  10: guava
  11: guyabano
  12: okra
  13: pea
  14: sayote

```

```convert_to_coco.py
#!/usr/bin/python
import os
import json
import xml.etree.ElementTree as ET

START_BOUNDING_BOX_ID = 1
PRE_DEFINE_CATEGORIES = {
    "ampalaya": 0,
    "atis": 1,
    "avocado": 2,
    "bell_pepper": 3,
    "calamansi": 4,
    "cucumber": 5,
    "dalandan": 6,
    "french_bean": 7,
    "green_chili": 8,
    "green_mango": 9,
    "guava": 10,
    "guyabano": 11,
    "okra": 12,
    "pea": 13,
    "sayote": 14
}

def get(root, name):
    return root.findall(name)

def get_and_check(root, name, length):
    vars = root.findall(name)
    if len(vars) == 0:
        raise NotImplementedError('Can not find %s in %s.'%(name, root.tag))
    if length > 0 and len(vars) != length:
        raise NotImplementedError('The size of %s is supposed to be %d, but is %d.'%(name, length, len(vars)))
    if length == 1:
        vars = vars[0]
    return vars

def convert(xml_dir, json_file):
    json_dict = {"images":[], "type": "instances", "annotations": [],
                 "categories": []}
    categories = PRE_DEFINE_CATEGORIES
    bnd_id = START_BOUNDING_BOX_ID
    image_id = 1

    # Get all XML files in directory
    xml_files = [f for f in os.listdir(xml_dir) if f.endswith('.xml')]
    
    for xml_file in xml_files:
        print(f"Processing {xml_file}")
        xml_path = os.path.join(xml_dir, xml_file)
        tree = ET.parse(xml_path)
        root = tree.getroot()

        size = get_and_check(root, 'size', 1)
        width = int(get_and_check(size, 'width', 1).text)
        height = int(get_and_check(size, 'height', 1).text)
        
        image = {
            'file_name': xml_file.replace('.xml', '.jpg'),
            'height': height,
            'width': width,
            'id': image_id
        }
        json_dict['images'].append(image)

        for obj in get(root, 'object'):
            category = get_and_check(obj, 'name', 1).text
            if category not in categories:
                continue
            category_id = categories[category]
            bndbox = get_and_check(obj, 'bndbox', 1)
            xmin = int(float(get_and_check(bndbox, 'xmin', 1).text))
            ymin = int(float(get_and_check(bndbox, 'ymin', 1).text))
            xmax = int(float(get_and_check(bndbox, 'xmax', 1).text))
            ymax = int(float(get_and_check(bndbox, 'ymax', 1).text))
            
            if xmax <= xmin or ymax <= ymin:
                continue
                
            o_width = abs(xmax - xmin)
            o_height = abs(ymax - ymin)
            ann = {
                'area': o_width*o_height,
                'iscrowd': 0,
                'image_id': image_id,
                'bbox':[xmin, ymin, o_width, o_height],
                'category_id': category_id,
                'id': bnd_id,
                'ignore': 0,
                'segmentation': []
            }
            json_dict['annotations'].append(ann)
            bnd_id += 1
        image_id += 1

    # Add categories
    for cate, cid in categories.items():
        cat = {'supercategory': 'none', 'id': cid, 'name': cate}
        json_dict['categories'].append(cat)

    # Save JSON file
    with open(json_file, 'w') as json_fp:
        json_str = json.dumps(json_dict, indent=4)
        json_fp.write(json_str)

def main():
    base_path = "E:/thesis models/RT-DETR-Pipeline/CamoCrops_2025"
    splits = ['Train', 'Validation', 'Test_new']
    
    for split in splits:
        xml_dir = os.path.join(base_path, 'Annotations', split)
        json_file = os.path.join(base_path, f'annotations_{split.lower()}.json')
        print(f"\nProcessing {split} split...")
        convert(xml_dir, json_file)
        print(f"Completed {split} conversion!")

if __name__ == '__main__':
    main()
```

i've also attached my annotation json files for your reference 

[annotations_test_new.json](https://github.com/user-attachments/files/18653792/annotations_test_new.json)
[annotations_train.json](https://github.com/user-attachments/files/18653793/annotations_train.json)
[annotations_validation.json](https://github.com/user-attachments/files/18653794/annotations_validation.json)

thank you in advance

**Comment by UltralyticsAssistant** [comment]
👋 Hello @MacJavellana, thank you for sharing such a detailed inquiry and for your interest in Ultralytics 🚀! We understand that working through dataset formatting and training nuances can be challenging, especially when transitioning between formats like VOC and COCO.

To effectively assist, here are a few pointers for debugging and resolving your issue:

1. **Dataset Formatting Issues**:
   Based on the error message, "No labels found," it seems there might be a mismatch in file paths or format definitions. Kindly double-check:  
   - That the paths specified in `camocrops.yaml` correctly correspond to your data's folder structure. If you notice any mismatches (e.g., `labels/annotations_test_new.json` vs. `label/sannotations_test_new.json`), please update them accordingly.  
   - That your COCO annotations include the correct `image_id` and are properly matched to your image files.

   Our [Dataset Formatting Guide](https://docs.ultralytics.com/datasets/) provides a comprehensive set of instructions to prepare your dataset for training. 😊

2. **Minimum Reproducible Example (MRE)**:
   If you still encounter this issue after verifying your dataset, could you provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/)? This would typically include:
   - A reduced version of your dataset (just a few images and annotations that replicate the issue).
   - A simplified version of your training script or key configurations.

   The MRE will help us diagnose your issue faster.

3. **Upgrade and Verify**:
   Ensure you have the latest version of the `ultralytics` package and confirm your environment is correctly configured with all requirements. Run the following to upgrade:  
   ```bash
   pip install -U ultralytics
   ```

4. **Community Support**:
   While you're exploring solutions, we encourage you to join our [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time support or continue discussions on our [Discourse](https://community.ultralytics.com/) or [Subreddit](https://reddit.com/r/Ultralytics).

### Environments

YOLO can be trained/tested in various environments with all dependencies preinstalled, such as [Google Colab](https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb), [Paperspace](https://console.paperspace.com/github/ultralytics/ultralytics), or [Kaggle](https://www.kaggle.com/models/ultralytics/yolo11). Feel free to leverage these for testing your setup.

<a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

### Status

You can also check the status of our [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) (<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>) to verify that all examples are functioning properly.

This is an automated response to guide you toward potential solutions. An Ultralytics engineer will review your issue shortly to provide more specific assistance. Stay tuned! 🚀

**Comment by MacJavellana** [comment]
hello... so i've fixed it i've followed this video specifically how they did their file structure where they have a folder called data with subfolders of images and labels

images has train, validate, and test
label has train, validate, and test (these has txt files instead of json)

here is the youtube link if anyone have the same problem in the future https://www.youtube.com/watch?v=m9fH9OWn8YM

**Reply by glenn-jocher** [reply]
Thank you for sharing your solution! For others encountering similar "No labels found" issues, ensure your dataset follows the YOLO format with images and labels in parallel directories as shown in our [Dataset Guide](https://docs.ultralytics.com/datasets). The key points are:
1. Verify labels exist in `../labels/train/` matching image filenames
2. Ensure normalized coordinates (0-1) in label TXT files
3. Double-check paths in your YAML configuration

Glad you resolved it! 🚀
